src/Raven.Server/Rachis/Follower.cs

```diff
@@ -411,7 +411,74 @@ private void NegotiateWithLeader(TransactionOperationContext context, LogLengthN
             // the reason we send this is to simplify the # of states in the protocol
 
             var snapshot = _connection.ReadInstallSnapshot(context);
-            _debugRecorder.Record("Snapshot received");
+            _debugRecorder.Record("Start receiving snapshot");
+
+            // reading the snapshot from network and committing it to the disk might take a long time. 
+            using (var cts = new CancellationTokenSource())
+            {
+                var readAndCommitSnapshotTask = Task.Run(() => ReadAndCommitSnapshot(context, snapshot, cts.Token), cts.Token);
+                try
+                {
+                    WaitForTaskCompletion(readAndCommitSnapshotTask, "ReadAndCommitSnapshot");
+                }
+                catch
+                {
+                    cts.Cancel();
+                    throw;
+                }
+            }
+
+            _debugRecorder.Record("Snapshot was received and committed");
+
+            // notify the state machine, we do this in an async manner, and start
+            // the operator in a separate thread to avoid timeouts while this is
+            // going on
+            var task = Task.Run(() => _engine.SnapshotInstalledAsync(snapshot.LastIncludedIndex));
+            WaitForTaskCompletion(task, "SnapshotInstalledAsync");
+
+            _debugRecorder.Record("Done with StateMachine.SnapshotInstalled");
+
+            // we might have moved from passive node, so we need to start the timeout clock
+            _engine.Timeout.Start(_engine.SwitchToCandidateStateOnTimeout);
+
+            _debugRecorder.Record("Snapshot installed");
+            //Here we send the LastIncludedIndex as our matched index even for the case where our lastCommitIndex is greater
+            //So we could validate that the entries sent by the leader are indeed the same as the ones we have.
+            _connection.Send(context, new InstallSnapshotResponse
+            {
+                Done = true,
+                CurrentTerm = _term,
+                LastLogIndex = snapshot.LastIncludedIndex
+            });
+
+            _engine.Timeout.Defer(_connection.Source);
+        }
+
+        private void WaitForTaskCompletion(Task task, string debug)
+        {
+            var sp = Stopwatch.StartNew();
+
+            var timeToWait = (int)(_engine.ElectionTimeout.TotalMilliseconds / 4);
+
+            using (_engine.ContextPool.AllocateOperationContext(out JsonOperationContext context))
```

Comment:
I think it might be a better idea to move the context into the while.
The reason is that we want to reset / renew it on each iteration, rather than risk it getting too big if we wait for a lot of time

Reply:
done